File: medium/0003. Longest Substring Without Repeating Characters.py
from collections import defaultdict


class Solution(object):
    def lengthOfLongestSubstring(self, s: str) -> int:
        letters = defaultdict(int)
        back = 0
        front = 0
        longest = 0
        while front < len(s):
            letters[s[front]] += 1
            while letters[s[front]] > 1:
                letters[s[back]] -= 1
                back += 1
            longest = max(longest, front - back + 1)
            front += 1
        return longest


# print(Solution().lengthOfLongestSubstring('abcabcbb'))

File: 2.medium/0003. Longest Substring Without Repeating Characters.cpp
// # Time Complexity:   O(n)
// # Memory Complexity: O(1)

#include <string>
#include <unordered_map>
#include <iostream>

using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, bool> charmap;
        int best = 1;
        int l = 0, r = 0;
        while (r < (int)s.length())
        {
            while (charmap.count(s[r]) && charmap[s[r]])
            {
                charmap[s[l]] = false;
                l++;
            }
            charmap[s[r]] = true;
            best = max(best, r - l + 1);
            r++;
        }
        return min(best, (int)s.length());
    }
};

// int main()
// {
//     Solution sol;
//     string s = "abcabcbb";
//     int result = sol.lengthOfLongestSubstring(s);
//     cout << result << endl;
//     s = "tmmzuxt";
//     result = sol.lengthOfLongestSubstring(s);
//     cout << result << endl;

//     return 0;
// }